### PR TITLE
Feature: DAC/ADC Diagnostics

### PR DESCRIPTION
**Pull request by dan-fern**
Prior to this pull request, `throttle_control_module.ino` had no in-line test to compare DAC output to ADC input as it is fedback in.  After this pull request, this test is added (along with a check_pedal_override) with a failure threshold of ± 50mV discrepancy between the DAC and ADC signals.  This threshold represents 1% of the possible voltage range, but is not empirically-driven and thus subject to future changes.  In addition to this test, the `dac_adc_diagnostics.ino` is modified to bench test some of the techniques implemented.

### CODE REVIEW

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 1]
```diff
@@ -60,7 +61,7 @@
 #define CAN_INIT_RETRY_DELAY (50)
 
 // ms
-#define PS_CTRL_RX_WARN_TIMEOUT (250)
+#define PS_CTRL_RX_WARN_TIMEOUT (2500)
```
Comment (LucasBuckland) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: This seems like a long time, was there some reasoning behind this value?

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 2]
```diff
@@ -257,49 +265,127 @@ void disableControl()
...
+    }
+    else
+    {
+        pedal_override = 0;
+    }
+}
+
+//
+void check_spoof_voltages(
+        bool first_ADC,
```
Comment (LucasBuckland) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: What is first_ADC? Is it still used?

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 3]
```diff
@@ -495,52 +596,48 @@ void loop()
...
-        local_override = 0;
+        incoming_serial_byte = Serial.read( );
+        process_serial_byte( );
     }
 
-    // read and parse incoming serial commands
-    if (Serial.available() > 0) {
-        incomingSerialByte = Serial.read();
-        processSerialByte();
+    if ( ++loop_counter % 10 == 0 )
```
Comment (LucasBuckland) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Could make this into a macro to be cleaner (and more descriptive), but being picky here really a suggestion.

**Comment by evanlivingston** [comment]
I'm closing this request until we've tested the voltage checks with the brake pump and steering motor.

**Comment by petebrink** [comment]
Lines 53 - 76: #defines should have parentheses around them
Lines 108 & 131: Why are there two functions that do basically the same thing?  Better to have one that is consistent
Lines 178 - 197: These variables should be defined earlier in the file. Also declaring multiple variables with one type definition is not allowed, e.g. int a, b; should be int a; int b;
Lines 207 & 241: Consolidate the averaging function into a separate function that both enable and disable control use.
Line 210: Rename AVG_max to something more meaningful like num_samples
Line 307: Change logic. there should only be a single return in a function.
Line 310 & 311: Do not use "int" as it is not clear what the width of an "int" is.
Line 317: Might want to create a structure to store the passed in values, e.g. struct spoof_data { int16_t high, int16_t low };
In general, when converting from integer types to floating types, you should be cautious about how the compiler will perform the conversion
Is the control sending values to the DAC all the time?
There might some problems with initialization in terms of when the control is enabled.
How do you determine when to run the check_spoof_voltages() function?
Line 358: It would be good to document these options in the function header.
Line 405: Brace formatting
Line 618: Increment loop counter outside of the if statement
I would declare loop_counter as a static local variable and not use % operator.
loop_counter is going to overflow quite often and will cause you to execute this check at odd intervals.  Better to reset the counter after each check
static int16_t loop_counter = 0;

if ( loop_counter < 10 )
{
    loop_counter += 1;
}
else
{
    loop_counter = 0;
    check_spoof_voltages()
}

**Comment by petebrink** [comment]
Lines 53 - 76: #defines should have parentheses around them
Lines 108 & 131: Why are there two functions that do basically the same thing?  Better to have one that is consistent
Lines 178 - 197: These variables should be defined earlier in the file. Also declaring multiple variables with one type definition is not allowed, e.g. int a, b; should be int a; int b;
Lines 207 & 241: Consolidate the averaging function into a separate function that both enable and disable control use.
Line 210: Rename AVG_max to something more meaningful like num_samples
Line 307: Change logic. there should only be a single return in a function.
Line 310 & 311: Do not use "int" as it is not clear what the width of an "int" is.
Line 317: Might want to create a structure to store the passed in values, e.g. struct spoof_data { int16_t high, int16_t low };
In general, when converting from integer types to floating types, you should be cautious about how the compiler will perform the conversion
Is the control sending values to the DAC all the time?
There might some problems with initialization in terms of when the control is enabled.
How do you determine when to run the check_spoof_voltages() function?
Line 358: It would be good to document these options in the function header.
Line 405: Brace formatting
Line 618: Increment loop counter outside of the if statement
I would declare loop_counter as a static local variable and not use % operator.
loop_counter is going to overflow quite often and will cause you to execute this check at odd intervals.  Better to reset the counter after each check
static int16_t loop_counter = 0;

if ( loop_counter < 10 )
{
    loop_counter += 1;
}
else
{
    loop_counter = 0;
    check_spoof_voltages()
}

**Pull request closed without merging**